history/src/postgres/postgres_transaction.rs

```diff
@@ -36,33 +37,49 @@ impl PostgresTransaction {
         }
     }
 
-    pub fn create_statement(schema: &String) -> String {
+    pub fn build_create_table_statement(epoch: EpochRef) -> String {
+        let schema = PostgresEpoch::build_schema_name(epoch);
+        format!(
+            r#"
+                CREATE TABLE IF NOT EXISTS {schema}.transactions (
+                    signature TEXT NOT NULL,
+                    slot BIGINT NOT NULL,
+                    err TEXT,
+                    cu_requested BIGINT,
+                    prioritization_fees BIGINT,
+                    cu_consumed BIGINT,
+                    recent_blockhash TEXT NOT NULL,
+                    message TEXT NOT NULL,
+                    CONSTRAINT pk_transaction_sig PRIMARY KEY(signature)
+                  );
+                  CREATE EXTENSION IF NOT EXISTS btree_gin;
+                  CREATE INDEX idx_slot ON {schema}.transactions USING btree (slot);
+                  CLUSTER {schema}.transactions USING idx_slot;
+            "#,
+            schema = schema
+        )
+    }
+
+    pub fn build_foreign_key_statement(epoch: EpochRef) -> String {
+        let schema = PostgresEpoch::build_schema_name(epoch);
         format!(
-            "\
-        CREATE TABLE {}.TRANSACTIONS (
-            signature CHAR(88) NOT NULL,
-            slot BIGINT, 
-            err STRING,
-            cu_requested BIGINT,
-            prioritization_fees BIGINT,
-            cu_consumed BIGINT,
-            recent_blockhash STRING NOT NULL,
-            message STRING NOT NULL,
-            PRIMARY KEY (signature)
-            CONSTRAINT fk_transactions FOREIGN KEY (slot) REFERENCES {}.BLOCKS(slot);
-          );
-        ",
-            schema, schema
+            r#"
+                ALTER TABLE {schema}.transactions
+                ADD CONSTRAINT fk_transactions FOREIGN KEY (slot) REFERENCES {schema}.blocks (slot);
+            "#,
+            schema = schema
         )
     }
 
-    pub async fn save_transactions(
+    pub async fn save_transaction_batch(
         postgres_session: &PostgresSession,
-        schema: &String,
+        epoch: EpochRef,
+        slot: Slot,
         transactions: &[Self],
     ) -> anyhow::Result<()> {
-        let mut args: Vec<&(dyn ToSql + Sync)> =
-            Vec::with_capacity(NB_ARUMENTS * transactions.len());
+        const NB_ARGUMENTS: usize = 8;
+        let tx_count = transactions.len();
+        let mut args: Vec<&(dyn ToSql + Sync)> = Vec::with_capacity(NB_ARGUMENTS * tx_count);
```

Comment:
A block can contain several thousand of Tx. This wait to process to build the query is not very efficient. In the whole process, there are several container allocations and transformation and in the end in the Postgres session everything is build into a string. Perhaps there's a way to avoid using a string for the query and to build in one pass.

Reply:
sure.that's not my code. will need to improve that. I know how but that's not my focus for this PR.

Reply:
Ok I agree. Perhaps we should open an issue to remember this evolution?